fgpyo/sam/__init__.py

```diff
@@ -584,3 +586,110 @@ def set_pair_info(r1: AlignedSegment, r2: AlignedSegment, proper_pair: bool = Tr
     insert_size = isize(r1, r2)
     r1.template_length = insert_size
     r2.template_length = -insert_size
+
+
+@attr.s(auto_attribs=True, frozen=True)
+class Template:
+    """A container for alignment records corresponding to a single sequenced template
+    or insert.
+
+    Attributes:
+        name: the name of the template/query
+        r1: Primary alignment for read 1, or None if there is none
+        r2: Primary alignment for read 2, or None if there is none
+        r1_supplementals: Supplementary alignments for read 1
+        r2_supplementals: Supplementary alignments for read 2
+        r1_secondaries: Secondary (non-primary) alignments for read 1
+        r2_secondaries: Secondary (non-primary) alignments for read 2
```

Comment:
do you want to do any validation on these?  E.g. must have either r1 or r2?  That all reads have the same query name as each other and `name`?  That `r1`, `r2` are primary, that `*_supplementals` are really supplementals, etc.?  You would be fine if you use the `build` method, but I could see bugs occur if someone uses `attr.evolve` and there's not validation.

Reply:
`name` could be a property

```python
@property
def name(self) -> str:
   return r1.query_name if r1 is not None else r2.query_name
```

Reply:
I'm 50/50 on the validation.  I don't necessarily want to incurr that overhead on every construction, especially if the TemplateIterator is doing it.  I don't think r1 or r2 has to be present ... the validations would boil down to validating flags on everything that _is_ present, _and_ validating that the query_name is the same across all records.  Thoughts?

`name` could be a property, but it require searching all reads/lists to find a record, so I think it's nicer to compute it when we're iterating all the reads at construction.  Do you disagree now that I've fixed the way it is collected?

Reply:
I see the value of not validating upon every construction.  I see two paths forward:

1. have a private attributed `_validated` that be set by the `TemplateIterator` to `true` to skip validation, and if not set, the validation methods are no longer no-ops?
2. make it clear in the docstring that validation does not occur?

I'd lean to (2) for now, but (1) is cute.

Reply:
I split the difference and:

1. Implemented a `validate()` function that asserts all kinds of things
2. Made the `build()` function call that, but with a parameter to disable it
3. Added class level docs to encourage use of `Template.build()` over `Template()`